pints/toy/_sho_model.py

```diff
@@ -0,0 +1,115 @@
+#
+# simple harmonic oscillator toy model.
+#
+# This file is part of PINTS.
+#  Copyright (c) 2017-2019, University of Oxford.
+#  For licensing information, see the LICENSE file distributed with the PINTS
+#  software package.
+#
+from __future__ import absolute_import, division
+from __future__ import print_function, unicode_literals
+import numpy as np
+import pints
+
+from . import ToyModel
+
+
+class SimpleHarmonicOscillatorModel(pints.ForwardModelS1, ToyModel):
+
+    r"""
+    Simple harmonic oscillator model for a particle that experiences a force
+    in proportion to its displacement from an equilibrium position,
+    and, in addition, a friction force. The system's behaviour is determined by
+    a second order ordinary differential equation (from Newton's second law):
+
+    .. math::
+        \frac{d^2y}{dt^2} = -y(t) - \theta \frac{dy(t)}{dt}
+
+    Here it has been assumed that the particle has unit mass and that the
+    restoring force has constant of proportionality equal to 1.
+
+    The model has three parameters: the initial position of the particle,
+    ``y(0)``, its initial momentum, ``dy/dt(0)`` and the magnitude of the
+    friction force, ``theta``.
+
+    Extends :class:`pints.ForwardModel`, :class:`pints.toy.ToyModel`.
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Simple_harmonic_motion
+    """
+    def __init__(self):
+        super(SimpleHarmonicOscillatorModel, self).__init__()
+
+    def n_parameters(self):
+        """ See :meth:`pints.ForwardModel.n_parameters()`. """
+        return 3
+
+    def simulate(self, parameters, times):
+        """ See :meth:`pints.ForwardModel.simulate()`. """
+        return self._simulate(parameters, times, False)
+
+    def simulateS1(self, parameters, times):
+        """ See :meth:`pints.ForwardModelS1.simulateS1()`. """
+        return self._simulate(parameters, times, True)
+
+    def _simulate(self, parameters, times, sensitivities):
+        y1_0, y2_0, theta = [float(x) for x in parameters]
+        times = np.asarray(times)
+        if np.any(times < 0):
+            raise ValueError('Negative times are not allowed.')
+
+        if theta != 2:  # i.e. not critically damped
+            sqrt = np.sqrt(theta**2 - 4 + 0j)
+            exp1 = np.exp(0.5 * times * (-theta - sqrt))
+            exp2 = np.exp(0.5 * times * (-theta + sqrt))
+
+            values = 1 / (2 * sqrt) * (
+                2 * y2_0 * (exp2 - exp1) + y1_0 * theta * (exp2 - exp1) +
+                y1_0 * sqrt * (exp1 + exp2)
+            )
+            values = np.real(values)
+
+            if sensitivities:
+                exp3 = np.exp(-0.5 * theta * times)
+                exp4 = np.exp(times * sqrt)
```

Comment:
Looks like we could cache some of these, but might not be worth the effort (and reduced legibility)!

Reply:
Yeah, not sure it's worth it since, whenever theta changes (as would occur in inference), they would need to be updated.